src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -369,38 +404,43 @@ public class SqlMutationEngine : IMutationEngine
                     throw _dabExceptionWithTransactionErrorMessage;
                 }
 
-                DbResultSetRow? dbResultSetRow = upsertOperationResult is not null ?
-                    (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
+                Dictionary<string, object?> resultRow = upsertOperationResultSetRow!.Columns;
 
-                if (upsertOperationResult is not null &&
-                    dbResultSetRow is not null && dbResultSetRow.Columns.Count > 0)
+                // For MsSql, MySql, if it's not the first result, the upsert resulted in an INSERT operation.
+                // Even if its first result, postgresql may still be an insert op here, if so, return CreatedResult
+                if (_sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL)
                 {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
-
-                    bool isUpdateResultSet = false;
-                    if (upsertOperationResult.ResultProperties.TryGetValue(IS_UPDATE_RESULT_SET, out object? isUpdateResultSetValue))
-                    {
-                        isUpdateResultSet = Convert.ToBoolean(isUpdateResultSetValue);
-                    }
+                    isUpdateResultSet = !PostgresQueryBuilder.IsInsert(resultRow);
+                }
 
-                    // For MsSql, MySql, if it's not the first result, the upsert resulted in an INSERT operation.
-                    // Even if its first result, postgresql may still be an insert op here, if so, return CreatedResult
-                    if (!isUpdateResultSet ||
-                        (_sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL &&
-                        PostgresQueryBuilder.IsInsert(resultRow)))
+                // When read permissions is configured without database policy, a subsequent select query will not be executed.
+                // However, the read action could have include and exclude fields configured and it
+                // Any additional fields that are present in the response are removed.
+                if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                {
+                    HashSet<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read).ToHashSet();
```

Comment:
```suggestion
                    IEnumerable<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read);
```

Reply:
The reason for using a hash set is to be able to perform O(1) lookups to determine whether a field in the response is one of the allowed columns or not.

Reply:
That will still be applicable with IEnumerable since hashset will have its overridden implementation of the method.

Reply:
Missed that `_authorizationResolver.GetAllowedExposedColumns()` returns as `HashSet`. So, can directly use `IEnumerable`